### DISCUSSION TITLE
Welcome to pyvirtualcam Discussions!

### DISCUSSION THREAD

**Discussion by letmaik** [letmaik/pyvirtualcam#12] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by FredericDlugi** [comment]
Hi,
This project really helped my project (developing a webcam driver for Basler/pylon-cameras). Since these cameras also use YUV, BGR and Mono Formats and the old obs-virtualcam implementation supported at least YUV I made these visible through the python-bindings. It seems like the focus of your project is only to support the RGB image format.
First: I was wondering, if there are any plans to support other pixel formats?
Second: Is there a specific reason, why the pixel format conversion is not done through an API like OpenCV but through the code in native_shared?

**Reply by letmaik** [reply]
Hey, glad to hear that you found the library useful.

The built-in virtual cameras of OBS on Windows and macOS only support a single image format, on Windows NV12, and macOS UYVY, both YUV variants. On Linux, via v4l2, probably more formats are natively supported, though currently I use UYVY there as well. See https://github.com/letmaik/pyvirtualcam/pull/25 on reasons for moving away from obs-virtualcam on Windows.

Currently there are no plans to support other pixel formats, mostly to keep it simple. If more people ask for it, it may happen.
The reason for not using OpenCV for format conversion is that it doesn't support converting from RGB to UYVY (it only has the opposite direction). Since those conversions are fairly simple it made sense to do it in native code, which also avoids an extra dependency. OpenCV is not a particularly lightweight library.

I saw that in your case you use YUYV and grayscale formats. If the library hypothetically would support YUYV as input, then there would have to be extra code to convert YUYV to NV12 (vertical downsampling + reordering) and YUYV to UYVY (trivial, just re-ordering) in order to support all platforms. Grayscale could be either done in two steps (first convert to RGB with numpy by [replicating](http://www.socouldanyone.com/2013/03/converting-grayscale-to-rgb-with-numpy.html) the channel, and then using the existing native RGB code), or by natively supporting grayscale to NV12/UYVY conversion, which would be slightly faster. I have a feeling using numpy would be preferable as it should be fast enough and would avoid extra native code, also considering that grayscale is likely a niche use case.

Did you do some experiments with the new version by converting to RGB first with OpenCV? Is that too slow? Do you lose too much image quality?
